botorch_community/acquisition/bll_thompson_sampling.py

```diff
@@ -0,0 +1,158 @@
+    ) -> torch.Tensor:
+        if self.discrete and X_cand is None:
+            raise ValueError("X_cand must be provided if `discrete` is True.")
+
+        if X_cand is not None and not self.discrete:
+            logger.warning(
+                "X_cand provided but self.discrete is False. Ignoring X_cand."
+            )
+
+        X_next = torch.empty(
+            num_samples, self.model.num_inputs, dtype=torch.float64, device=self.device
+        )
+
+        # get max of sampled functions at candidate points for each function
+        for i in range(num_samples):
+            f = self.model.sample()
+
+            # NOTE: If self.discrete is False, we will always numerically optimize the sample
+            # path for X_next. X_cand is ignored in this case.
+            if not self.discrete:
+                # optimize sample path
+                X_cand, Y_cand = _optimize_sample_path(
+                    f, self.num_restarts, self.bounds, self.lb, self.ub, self.device
+                )
+            else:
+                # evaluate sample path at candidate points and select best
+                Y_cand = f(X_cand)
+
+            # select the best candidate
+            X_next[i, :] = X_cand[Y_cand.argmax()]
+
+        # ensure that the next point is within the bounds,
+        # scipy minimize can sometimes return points outside the bounds
+        X_next = torch.clamp(X_next, self.lb.to(self.device), self.ub.to(self.device))
+        return X_next
+
+
+def _optimize_sample_path(
+    f: torch.nn.Module,
+    num_restarts: int,
+    bounds: list[tuple[float, float]],
+    lb: torch.Tensor,
+    ub: torch.Tensor,
+    device: torch.device,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Helper function to optimize the sample path of a BLL network.
+
+    Args:
+        f: The sample to optimize.
+        num_restarts: Number of restarts for optimization-based sampling.
+        bounds: List of tuples specifying the lower and upper bounds for each input dimension.
+        lb: Lower bounds for each input dimension.
+        ub: Upper bounds for each input dimension.
+        device: Device on which to store the candidate points.
+
+    Returns:
+        Candidate points and corresponding function values.
+    """
+    X_cand = torch.empty(num_restarts, f.num_inputs, dtype=torch.float64, device=device)
+    Y_cand = torch.empty(
+        num_restarts, f.num_outputs, dtype=torch.float64, device=device
+    )
+
+    # create numpy wrapper around the sampled function, note we aim to maximize
+    def func(x):
+        return -f(torch.from_numpy(x).to(device)).detach().cpu().numpy()
+
+    # get gradient and create wrapper
+    grad_f = grad(lambda x: f(x).mean())
+
+    def grad_func(x):
+        return -grad_f(torch.from_numpy(x).to(device)).detach().cpu().numpy()
+
+    for j in range(num_restarts):
+        # generate initial condition from within the bounds
+        x0 = np.random.rand(f.num_inputs)
+        x0 = lb + (ub - lb) * x0
```

Comment:
Nit: Could do this once in a vectorized fashion outside the loop and then just index in the loop (more a µ-optimization though)